PumaRhino/utils.cpp

```diff
@@ -310,4 +317,164 @@ void PathRemover::operator()(std::filesystem::path const* p) {
+		found = i_str.find(i_delim, startIndex);
+	}
+	if (startIndex != i_str.size())
+		result.push_back(std::wstring(i_str.begin() + startIndex, i_str.end()).c_str());
+	return result;
+}
+
+std::vector<const wchar_t*> fromCeArray(const std::wstring& stringArray) {
+	return pcu::split(stringArray, CE_ARRAY_DELIMITER);
+}
+
+const std::wstring toCeArray(const wchar_t* const* values, size_t count) {
+	std::wstring serializedArray;
+	for (size_t i = 0; i < count; ++i) {
+		serializedArray += std::wstring(values[i]) + CE_ARRAY_DELIMITER;
+	}
+	return serializedArray;
+}
+
+const std::wstring toCeArray(const bool* values, size_t count) {
+	std::wstring serializedArray;
+	for (size_t i = 0; i < count; ++i) {
+		serializedArray += std::wstring(values[i] ? L"true" : L"false") + CE_ARRAY_DELIMITER;
+	}
+	return serializedArray;
+}
+
+const std::wstring toCeArray(const double* values, size_t count) {
+	std::wstring serializedArray;
+	for (size_t i = 0; i < count; ++i) {
+		serializedArray +=  std::to_wstring(values[i]) + CE_ARRAY_DELIMITER;
+	}
+	return serializedArray;
+}
+
+const std::wstring toCeArray(const int* values, size_t count) {
+	std::wstring serializedArray;
+	for (size_t i = 0; i < count; ++i) {
+		serializedArray +=  std::to_wstring(values[i]) + CE_ARRAY_DELIMITER;
+	}
+	return serializedArray;
+}
+
+/**
+ * Interop helpers
+ */
+
+template <typename T>
+void fillMapBuilder(const std::wstring& /* key */, T /* value */, AttributeMapBuilderPtr& /* aBuilder */) {
+	throw std::invalid_argument("Received type is not supported");
+}
+
+template <>
+void fillMapBuilder<int>(const std::wstring& key, int value, AttributeMapBuilderPtr& aBuilder) {
+	aBuilder->setBool(key.c_str(), static_cast<bool>(value));
+}
+
+template <>
+void fillMapBuilder<double>(const std::wstring& key, double value, AttributeMapBuilderPtr& aBuilder) {
+	aBuilder->setFloat(key.c_str(), value);
+}
+
+template <typename T>
+void fillArrayMapBuilder(const std::wstring& /* key */, const std::vector<const wchar_t*>& /* values */,
+                         AttributeMapBuilderPtr& /* aBuilder */) {
+	throw std::invalid_argument("Received type is not supported");
+}
+
+template <>
+void fillArrayMapBuilder<bool>(const std::wstring& key, const std::vector<const wchar_t*>& values,
+                               AttributeMapBuilderPtr& aBuilder) {
+	bool* bArray = new bool[values.size()];
+	for (int i = 0; i < values.size(); ++i) {
+		bArray[i] = (bool)(std::wstring(values[i]) == L"true");
+	}
+	aBuilder->setBoolArray(key.c_str(), bArray, values.size());
+}
+
+template <>
+void fillArrayMapBuilder<double>(const std::wstring& key, const std::vector<const wchar_t*>& values,
+                                 AttributeMapBuilderPtr& aBuilder) {
+	double* dArray = new double[values.size()];
+	for (int i = 0; i < values.size(); ++i) {
+		dArray[i] = (double)std::stod(std::wstring(values[i]));
+	}
+	aBuilder->setFloatArray(key.c_str(), dArray, values.size());
+}
```

Comment:
template functions should be moved to the header file so they are fully visible to the compiler at their instantiation site.